src/platform/platform_worker.c

```diff
@@ -76,49 +62,74 @@ typedef struct QUIC_CACHEALIGN CXPLAT_WORKER {
     //
     CXPLAT_SLIST_ENTRY* ExecutionContexts;
 
+    //
+    // The ideal processor for the worker thread.
+    //
+    uint16_t IdealProcessor;
+
+    //
+    // Flags to indicate what has been initialized.
+    //
+    BOOLEAN InitializedEventQ : 1;
+#ifdef CXPLAT_SQE_INIT
+    BOOLEAN InitializedShutdownSqe : 1;
+    BOOLEAN InitializedWakeSqe : 1;
+    BOOLEAN InitializedUpdatePollSqe : 1;
+#endif
+    BOOLEAN InitializedThread : 1;
+    BOOLEAN InitializedECLock : 1;
+
 } CXPLAT_WORKER;
 
+CXPLAT_LOCK CxPlatWorkerLock;
+CXPLAT_RUNDOWN_REF CxPlatWorkerRundown;
 uint32_t CxPlatWorkerCount;
 CXPLAT_WORKER* CxPlatWorkers;
 CXPLAT_THREAD_CALLBACK(CxPlatWorkerThread, Context);
 
 void
-CxPlatWorkerWake(
-    _In_ CXPLAT_WORKER* Worker
-    )
-{
-    CxPlatEventQEnqueue(&Worker->EventQ, &Worker->WakeSqe, (void*)&WorkerWakeEventPayload);
-}
-
-CXPLAT_EVENTQ*
-CxPlatWorkerGetEventQ(
-    _In_ uint16_t IdealProcessor
+CxPlatWorkersInit(
+    void
     )
 {
-    return &CxPlatWorkers[IdealProcessor % CxPlatWorkerCount].EventQ;
+    CxPlatLockInitialize(&CxPlatWorkerLock);
 }
 
 #pragma warning(push)
 #pragma warning(disable:6385)
 #pragma warning(disable:6386) // SAL is confused about the worker size
 BOOLEAN
-CxPlatWorkersInit(
-    void
+CxPlatWorkersLazyStart(
+    _In_opt_ QUIC_EXECUTION_CONFIG* Config
     )
 {
-    CxPlatWorkerCount = CxPlatProcActiveCount(); // TODO - use max instead?
+    CxPlatLockAcquire(&CxPlatWorkerLock);
```

Comment:
>CxPlatWorkerLock [](http://example.com/codeflow?start=23&length=16)

Is this lock necessary? It's only used in CxPlatWorkersLazyStart. I can't see how CxPlatWorkersLazyStart can be called from different threads.

Reply:
The platform layer is treated as a semi-public API, that could possibly be used my multiple consumers. Indeed, if you compile our tools with static linking, then the platform does end up getting shared by MsQuic and the tool itself. So just to be extra safe, we have synchronization.